### PR TITLE
Doc for layers and loading

### PR DESCRIPTION
**Pull request by TheBB**
I was encouraged to make a PR for this, so I rewrote it a little in a way that I feel it fits better.

Technically this file explains two things that are related but different
- how loading works (in Spacemacs as a whole, and Emacs)
- rules of thumb for how to write good layers

The structured writer in me would like to separate them more, but since we get so many questions about precisely these things I feel they should go together anyhow. It is of course not our job to explain Emacs internals, but I feel some overlap between our docs and the Emacs docs can go a long way.

### CODE REVIEW

**Comment by TheBB** [comment]
How do I get that cute keyboard markup in tildes?

**Comment by syl20bnr** [comment]
This is done here: https://github.com/syl20bnr/spacemacs/blob/develop/core/core-funcs.el#L112-L144

File: doc/LAYERS.org [hunk 1]
```diff
@@ -0,0 +1,530 @@
...
+* How do I ... idiomatically?
+
+** Setup auto-completion for a major mode
+In your layer's =config.el=, call =spacemacs|defvar-company-backends=.
+
+#+begin_src emacs-lisp
+  (spacemacs|defvar-company-backends yoyo-mode)
+#+end_src
+
+This creates a variable called =company-backends-python-mode=. In the package
```
Comment (person808) [inline doc/LAYERS.org]: Should be `yoyo-mode` not `python-mode` to match

**Comment by person808** [comment]
`VIMUSERS.org` also has some sections on layers so you may want to take a look at it and add a link to `LAYERS.org`.

**Comment by TheBB** [comment]
Thanks, updated. @syl20bnr I meant the Github rendered page, but I see now that none of the other pages have them. Was I hallucinating that?

**Comment by syl20bnr** [comment]
It was the case back in the time when docs were in markdown ;-)

**Comment by swaroopch** [comment]
+1, enlightening documentation.

**Comment by TheBB** [comment]
I guess we should keep this open for a while. When the dust has settled from recent developments I'll update.

File: doc/LAYERS.org [hunk 2]
```diff
@@ -0,0 +1,540 @@
...
+
+Use-package supports heaps of useful keywords. Look at the [[https://github.com/jwiegley/use-package][documentation]] for
+more.
+
+* Anatomy of a layer
+A layer is simply a folder somewhere in Spacemacs' layer search path that
+contains /at least one/ of these files, ordered vaguely by importance.
+
+- =packages.el= :: for specifying which packages the layer uses
+- =extensions.el= :: for specifying extensions
```
Comment (syl20bnr) [inline doc/LAYERS.org]: You can remove all the mentions about extensions.

File: doc/LAYERS.org [hunk 3]
```diff
@@ -0,0 +1,540 @@
...
+- =config.el= :: for miscellaneous configuration
+- =funcs.el= :: for defining general utility functions
+- =keybindings.el= :: for defining general keybindings
+
+Additionally, for each extension, there should be a folder
+=<layer>/extensions/<extension>/= containing the source code for that extension.
```
Comment (syl20bnr) [inline doc/LAYERS.org]: The folder will be renamed to "local" instead of "extensions". The new name will become mandatory in 0.105.0 (in about 2 months).

File: doc/LAYERS.org [hunk 4]
```diff
@@ -0,0 +1,540 @@
...
+Before initializing an extension, Spacemacs will add this folder to the load
+path for you, so that you can treat it as any other package. The only reason
+Spacemacs treats extensions differently is so that it can do this.
+
+** packages.el
+This file should define a variable called =<layer>-packages=, which should be a
+list of all the packages that this layer needs. Additionally, a layer may
+/exclude/ packages through the variable =<layer>-excluded-packages=. This will
```
Comment (syl20bnr) [inline doc/LAYERS.org]: `<layer>-excluded-packages` will be deprecated in 0.105.0, it is replaced by the property `:excluded` in the package list.

File: doc/LAYERS.org [hunk 5]
```diff
@@ -0,0 +1,540 @@
...
+A layer is simply a folder somewhere in Spacemacs' layer search path that
+contains /at least one/ of these files, ordered vaguely by importance.
+
+- =packages.el= :: for specifying which packages the layer uses
+- =extensions.el= :: for specifying extensions
+- =config.el= :: for miscellaneous configuration
+- =funcs.el= :: for defining general utility functions
+- =keybindings.el= :: for defining general keybindings
+
+Additionally, for each extension, there should be a folder
```
Comment (syl20bnr) [inline doc/LAYERS.org]: `extensions` will be referenced as `local packages` starting 0.104.0.

File: doc/LAYERS.org [hunk 6]
```diff
@@ -0,0 +1,540 @@
...
+/exclude/ packages through the variable =<layer>-excluded-packages=. This will
+block packages (and, despite the name, extensions) from being installed, even if
+they are included by another layer.
+
+For each included package, you may define one or more of the following
+functions, which are called in order by Spacemacs to initialize the package.
+
+1. =<layer>/pre-init-<package>=
+2. =<layer>/init-<package>=
+3. =<layer>/post-init-<package>=
```
Comment (syl20bnr) [inline doc/LAYERS.org]: To avoid confusion between spacemacs `init` functions and `init` from `use-package` and `use-package` hooks (`:pre-init` and `:post-init`). All `init` function will be renamed to `configure`. Will become mandatory in 0.105.0 as well.

File: doc/LAYERS.org [hunk 7]
```diff
@@ -0,0 +1,540 @@
...
+
+It is the responsibility of these functions to load and configure the package in
+question. Spacemacs will do nothing other than download the package and place it
+in the load path for you.
+
+*Note:* A package will not be installed unless at least one layer defines an
+=init= function for it. That is to say, in a certain sense, the =init= function
```
Comment (syl20bnr) [inline doc/LAYERS.org]: Now only one layer can define the `configure` function for a package. This layer is called the owner of the package. In the case of multiple `configure` function, the last encountered one is declared the owner.

File: doc/LAYERS.org [hunk 8]
```diff
@@ -0,0 +1,540 @@
...
+does mandatory setup while the =pre-init= and =post-init= functions do optional
+setup. This can be used for managing cross-layer dependencies, which we will
+discuss later.
+
+** extensions.el
```
Comment (syl20bnr) [inline doc/LAYERS.org]: This section is not needed anymore. Should be replaced by a section on a local packages which correspond to packages with the property `:location local`.

File: doc/LAYERS.org [hunk 9]
```diff
@@ -0,0 +1,540 @@
...
+
+More on this in the next section.
+
+* The Spacemacs loading process
+The Spacemacs loading process can be summarized as follows:
+
+1. A handful packages are installed and loaded as part of the bootstrap process.
+   Among these packages, of particular interest to us, is =use-package=. All
+   subsequent code in the loading process may make use of it.
+2. Spacemacs goes through all the enabled layers and evaluates their files
```
Comment (syl20bnr) [inline doc/LAYERS.org]: goes through all used layers in order of declaration. So it is possible to use this property to override a `configure` function and steal the ownership of a package in a custom layer.

File: doc/LAYERS.org [hunk 10]
```diff
@@ -0,0 +1,540 @@
...
+   and extension are executed in turn.
+
+It is step five that interests us. It is very important that a package is not
+installed if no =init= function is defined for it.
+
+We say that a layer *owns* a package if it defines an =init= function for it. A
+layer does *not* own a package if it only defines =pre-init= or =post-init=
+functions.
+
+Best practices dictate that only /one/ layer should own any given package.
```
Comment (syl20bnr) [inline doc/LAYERS.org]: Now effectively only one layer can take the ownership of a package.

File: doc/LAYERS.org [hunk 11]
```diff
@@ -0,0 +1,540 @@
...
+an arbitrary set of packages, there is no /a priori/ safe way to assume that
+another package is included. Use =configuration-layer/package-usedp= if you
+must.
+
+This can be very challenging, so please take this as a guideline and not
+something that is absolute. It is quite possible for the user to break her
+Spacemacs installation by excluding the wrong packages, and it is not our
+intention to prevent this at all costs.
+
+*** Don't assume a fixed order
```
Comment (syl20bnr) [inline doc/LAYERS.org]: Actually packages are loaded in alphabetical order so it is always possible to know in advance if A will be configured before B.

Comment (TheBB) [inline doc/LAYERS.org]: But is that behaviour which is documented and guaranteed, or is it simply undefined behaviour that happens to be this way?

**Comment by TheBB** [comment]
Wow, thanks a lot @syl20bnr. Saves me from having to figure it out. Hopefully I can get to this tonight.

File: doc/LAYERS.org (hunk 9 above)
Comment (TheBB) [inline doc/LAYERS.org]: The `spacemacs` layer goes first, then?

**Comment by TheBB** [comment]
Updated.

**Comment by TheBB** [comment]
Github rendering can't deal with footnotes. Is that a deal breaker?

**Comment by syl20bnr** [comment]
@TheBB I can live with it :-)

In the end I will keep `init` instead of `configure` for the ... `init` function. Sorry for the wrong indication.

My motivation about renaming the init functions (all of them, not only `init`, but also the `pre` and `post` variants) was to reduce the possible confusion between spacemacs init functions and use-package `:init` keyword. Turns out that `configure` is as confusing since it is like `:config`. Moreover according to use-package terminology, init functions are executed right away at initialization time so it would not make sense to rename the functions to `configure`. Last, it put too much burden on spacemacs users.

For these reasons I won't rename the init functions.

**Comment by TheBB** [comment]
Okay :-P

Updated and rebased. (It said it couldn't be automatically merged... idk why, the rebase was automatic.)

**Comment by TheBB** [comment]
Updated with `recipe`.

**Comment by TheBB** [comment]
I just shat on the bed with this. I'll fix it...

**Comment by TheBB** [comment]
Okay, we're back in business. Updated with the latest developments.

**Comment by TheBB** [comment]
I added a note about the new layer dependency functionality.

File: doc/DOCUMENTATION.org [hunk 12]
```diff
@@ -310,12 +313,13 @@ Where:
 | packages.el    | The list of packages to install and the functions to initialize them |
 
 =Packages= are =ELPA= packages which can be installed from an =ELPA= compliant
-repository, and =Extensions= are generally elisp code from git submodules.
+repository, and =Extensions= are generally elisp code from git submodules, or
+other code one might want to use but which isn't available on an =ELPA=
+compliant repository, for whatever reason.
```
Comment (syl20bnr) [inline doc/DOCUMENTATION.org]: This section is out of date. I think we can remove it. The difference between packages will be described somewhere else.

The file tree of a layer must be updated. There is no `extensions.el` anymore and `extensions` folder is now `local`.

Although these changes are out of scope of this PR I'm OK to make them here.

Comment (syl20bnr) [inline doc/DOCUMENTATION.org]: Or scratch that, I or another contributor will open a new PR to update all this stuff in DOCUMENTATION.org

File: doc/LAYERS.org [hunk 13]
```diff
@@ -0,0 +1,588 @@
...
+their first configuration layer, whether for private use or for contributing
+upstream. It should help clear up some confusion regarding how layers work and
+how Spacemacs (and Emacs) loads packages.
+
+* Nomenclature
+Layers and packages. What gives?
+
+- Package :: A set of Emacs Lisp files that, taken together, provide some
+     feature. Packages may be available on a package repository, such as ELPA or
+     MELPA.
```
Comment (syl20bnr) [inline doc/LAYERS.org]: as well as a service provider like github or even local.

File: doc/LAYERS.org [hunk 14]
```diff
@@ -0,0 +1,588 @@
...
+
+#+begin_src emacs-lisp
+  (eval-after-load 'helm
+    '(progn
+       ;; Code
+       ))
+#+end_src
+
+This arranges for the relevant code to be executed after Helm is loaded (using
+eithe =require= or an autoload), or if Helm is already loaded, the code is
```
Comment (syl20bnr) [inline doc/LAYERS.org]: typo eithe -> either

File: doc/LAYERS.org [hunk 15]
```diff
@@ -0,0 +1,588 @@
...
+
+#+begin_src emacs-lisp
+  (with-eval-after-load 'helm
+    ;; Code
+    )
+#+end_src
+
+** Use-package
+For /end users/ who are trying to put together an efficient Emacs configuration,
+there is a very useful /package/ called use-package that provides a macro called
```
Comment (syl20bnr) [inline doc/LAYERS.org]: you can emphasis use-package -> `=use=package=`

File: doc/LAYERS.org [hunk 16]
```diff
@@ -0,0 +1,588 @@
...
+
+- =packages.el= :: for specifying which packages the layer uses
+- =config.el= :: for miscellaneous configuration
+- =funcs.el= :: for defining general utility functions
+- =keybindings.el= :: for defining general keybindings
+
+Additionally, for each local package (see the next section), there should be a
+folder =<layer>/local/<package>/= containing the source code for that package.
+Before initializing that package, Spacemacs will add this folder to the load
+path for you, so that you can treat it as any other package from MELPA.
```
Comment (syl20bnr) [inline doc/LAYERS.org]: I would remove `, so that you can treat it as any other package from MELPA`.

File: doc/LAYERS.org (hunk 7 above)
Comment (syl20bnr) [inline doc/LAYERS.org]: Still need to be fixed.

Comment (syl20bnr) [inline doc/LAYERS.org]: In fact you are right, the at least one is correct.
I guess we need a paragraph on the package ownership.

File: doc/LAYERS.org [hunk 17]
```diff
@@ -0,0 +1,588 @@
...
+3. Spacemacs checks which packages should be downloaded and installed. To be
+   installed, a package must be
+   - included by a layer that the user has enabled,
+   - not be excluded by any other layer that the user has enabled,
+   - not be excluded by the user herself, and
+   - there must be at least one =<layer>/init-<package>= function defined for
+     it.
+   Alternatively, if a package is part of the end user's
+   =dotspacemacs-additional-packages=, it will also be installed.
+4. All packages which should be installed are, and all packages which are
```
Comment (syl20bnr) [inline doc/LAYERS.org]: I think `used by the user` is missing after `are`.

File: doc/LAYERS.org (hunk 7 above)
Comment (syl20bnr) [inline doc/LAYERS.org]: Scrap that........... I should read the whole article before commenting ;-D

File: doc/LAYERS.org [hunk 18]
```diff
@@ -0,0 +1,588 @@
...
+- =configuration-layer/layer-usedp= :: check if a layer is enabled
+- =configuration-layer/package-usedp= :: check if a package is or will be installed
+
+These are useful in some cases, but usually you can get the desired result just
+by using =post-init= functions.
+
+For layers that require another layers to be enabled, use the functions
+=configuration-layer/declare-layer= and =configuration-layer/declare-layers= to
+ensure that layers are enabled even if the user has not enabled them explicitly.
+Calls to these functions should go in the =config.el= file.
```
Comment (syl20bnr) [inline doc/LAYERS.org]: should -> must :-)

File: doc/LAYERS.org [hunk 19]
```diff
@@ -0,0 +1,588 @@
...
+    )
+#+end_src
+
+Since a call to =use-package= may evaluate the =:init= block immediately, any
+function that wants to inject code into this block must run =before= the call to
+=use-package=. Further, since this call to =use-package= typically takes place
+in the =init-<package>= function, calls to =spacemacs|use-package-add-hook=
+typically happen in the =pre-init-<package>= functions, and not in
+=post-init-<package>=.
+
```
Comment (syl20bnr) [inline doc/LAYERS.org]: You can add that, as a rule of thumb, putting use-package hooks in `pre-init` is pretty safe so it should be the default choice when hooking stuff to use-package.

File: doc/LAYERS.org (hunk 11 above)
Comment (syl20bnr) [inline doc/LAYERS.org]: packages are explicitly sorted before being configured.

File: doc/LAYERS.org [hunk 20]
```diff
@@ -0,0 +1,588 @@
...
+Spacemacs installation by excluding the wrong packages, and it is not our
+intention to prevent this at all costs.
+
+*** Don't assume a fixed order
+In Spacemacs, layers are loaded in order of inclusion in the dotfile, and
+packages are loaded in alphabetical order. You should not depend on the latter,
+and in most cases you should not depend on the former, either. Preferably, write
+your layer so that it is independent of load ordering. The =pre=- and
+=post-init= functions are helpful, together with =configuration-layer/package-usedp=.
+
```
Comment (syl20bnr) [inline doc/LAYERS.org]: You are right about this since stuff can be lazy loaded but it is guaranteed that packages are configured alphabetically so when you know what you are doing you can rely on this property. It is the cases at several places in Spacemacs.

But maybe we should not document this to keep things simple. Relying on configuration order may be considered to be some kind of optimization we can do post merge of a PR.

**Comment by syl20bnr** [comment]
I love this doc :+1:

File: doc/LAYERS.org (hunk 17 above)
Comment (TheBB) [inline doc/LAYERS.org]: I meant *All packages which should be installed are installed*, but I guess I can write that explicitly.

**Comment by TheBB** [comment]
Thanks, @syl20bnr. I have rebased and updated. I left the part about load order intact with a warning that it shouldn't be relied upon unless necessary.

Hopefully this is the last iteration of this PR. :-)

**Comment by syl20bnr** [comment]
It's OK for me, nice work :+1:

**Comment by syl20bnr** [comment]
This wonderful doc is now merged in develop.
Thank you! :heart:
